### PR TITLE
loadbalancingexporter: narrow down critical area of ConsumeTraces

### PR DESCRIPTION
**Pull request by ZhengHe-MD**
**Description:**

quick fix on the problem that rolling updates of next tier collectors will stop dns resolver from periodical checking.

**Link to tracking Issue:** 

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1690

**Testing:**

Add a test simulate rolling updates.

**Documentation:**

None

### CODE REVIEW

File: exporter/loadbalancingexporter/exporter_test.go [hunk 1]
```diff
@@ -521,6 +524,110 @@ func TestNoTracesInBatch(t *testing.T) {
...
+		for {
+			select {
+			case <- ticker.C:
+				go p.ConsumeTraces(context.Background(), randomTraces())
+			}
+		}
+	}()
+
+	// wait until rolling updates finish
+	time.Sleep(100 * time.Millisecond)
```
Comment (jpkrohling) [inline exporter/loadbalancingexporter/exporter_test.go:615]: Instead of waiting a specific time, which might lead to intermittent CI failures, it would be better to create a WaitGroup and Add/Wait based on conditions.

File: exporter/loadbalancingexporter/exporter_test.go [hunk 2]
```diff
@@ -521,6 +524,110 @@ func TestNoTracesInBatch(t *testing.T) {
...
+	require.NoError(t, err)
+	defer p.Shutdown(context.Background())
+	// ensure using default exporters
+	p.exporters = defaultExporters
+	p.res.onChange(func(endpoints []string) {
+		p.exporters = defaultExporters
+	})
+
+	// keep consuming traces every 2ms
+	go func() {
```
Comment (jpkrohling) [inline exporter/loadbalancingexporter/exporter_test.go:604]: Even if it's just a test, you'd need a way to break out of this loop. Perhaps add a channel that is closed after the test is finished (currently, the `time.Sleep` on line 615)?

File: exporter/loadbalancingexporter/exporter_test.go [hunk 3]
```diff
@@ -521,6 +524,110 @@ func TestNoTracesInBatch(t *testing.T) {
...
+		ticker := time.NewTicker(2 * time.Millisecond)
+		for {
+			select {
+			case <- ticker.C:
+				go p.ConsumeTraces(context.Background(), randomTraces())
```
Comment (jpkrohling) [inline exporter/loadbalancingexporter/exporter_test.go:609]: Do you need this to be another go routine? With this change, this shouldn't block long anymore, right?

**Review by jpkrohling — COMMENTED** [review]
The fix looks great, the test needs some attention.

File: exporter/loadbalancingexporter/exporter_test.go (hunk 3 above)
Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go:609]: yes, you're right. I should remove the `go`.

**Comment by ZhengHe-MD** [comment]
@jpkrohling thx for reviewing my codes. I will certainly pay attention to the test. and tell you when it's ready.

File: exporter/loadbalancingexporter/exporter_test.go (hunk 3 above)
Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go:609]: ah. after a deep thought, I found that it does need the go. because I use time.Sleep in ConsumeTraces to simulate unreachable backends, if I don't use go routine, this loop will stuck there and stop consuming traces.

**Comment by ZhengHe-MD** [comment]
@jpkrohling the PR is ready for review, please take a look when you have time

File: exporter/loadbalancingexporter/exporter_test.go [hunk 4]
```diff
@@ -521,6 +525,131 @@ func TestNoTracesInBatch(t *testing.T) {
...
+
+	// give limited but enough time to rolling updates. otherwise this test
+	// will still pass due to the 10 secs of sleep that is used to simulate
+	// unreachable backends.
+	go func() {
+		time.Sleep(50 * time.Millisecond)
+		resolverCh <- 1
+	}()
+
+	<-resolverCh
```
[the remark this answers is not in the document]
Comment (pkositsyn) [reply exporter/loadbalancingexporter/exporter_test.go:635]: I see now, that it is done to avoid blocking in tests. Then I would just suggest making a buffered channel (size 1 is enough) to avoid goroutine leak

[the remark this answers is not in the document]
Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go:635]: thx, `chan int` does look confusing, will use `chan struct{}`, and also a size 1 buffered channel

File: exporter/loadbalancingexporter/exporter_test.go [hunk 5]
```diff
@@ -521,6 +525,131 @@ func TestNoTracesInBatch(t *testing.T) {
...
+	defer p.Shutdown(context.Background())
+	// ensure using default exporters
+	p.updateLock.Lock()
+	p.exporters = defaultExporters
+	p.updateLock.Unlock()
+	p.res.onChange(func(endpoints []string) {
+		p.updateLock.Lock()
+		p.exporters = defaultExporters
+		p.updateLock.Unlock()
+	})
```
[the remark this answers is not in the document]
Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go:609]: yes! you're right! and I just know this after you pointed out. but that's a hidden fact for this specific test case and I would rather keep this for the sake of readability. if I change the test case to simulate the following process:

```
["127.0.0.1", "127.0.0.2"] ->
["127.0.0.1", "127.0.0.2", "127.0.0.3"] -> 
["127.0.0.2", "127.0.0.3"] -> 
["127.0.0.2", "127.0.0.3", "127.0.0.4"] ->
["127.0.0.3", "127.0.0.4"]
```

the `onChange` can't be a no-op function then.

**Comment by ZhengHe-MD** [comment]
@pkositsyn thx for reviewing this, and I've modify the test according to your suggestions. please see if there is any other issue when you have time. 😬

File: exporter/loadbalancingexporter/exporter_test.go [hunk 6]
```diff
@@ -521,6 +525,134 @@ func TestNoTracesInBatch(t *testing.T) {
...
+				return
+			case <-ticker.C:
+				go p.ConsumeTraces(ctx, randomTraces())
+			}
+		}
+	}(ctx)
+
+	// give limited but enough time to rolling updates. otherwise this test
+	// will still pass due to the 10 secs of sleep that is used to simulate
+	// unreachable backends.
```
Comment (jpkrohling) [inline exporter/loadbalancingexporter/exporter_test.go:632]: Is this related to the condition on line 556? If so, you can add a WaitGroup there and get rid of guessing timings.

File: exporter/loadbalancingexporter/exporter_test.go [hunk 7]
```diff
@@ -521,6 +525,134 @@ func TestNoTracesInBatch(t *testing.T) {
 	}
 }
 
+func TestRollingUpdatesWhenConsumeTraces(t *testing.T) {
+	// this test is based on the discussion in the following issue for this exporter:
+	// https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1690
+	// prepare
+
+	// simulate rolling updates, the dns resolver should resolve in the following order
+	// ["127.0.0.1"] -> ["127.0.0.1", "127.0.0.2"] -> ["127.0.0.1"]
```
[the remark this answers is not in the document]
Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go:534]: my fault, thx!

File: exporter/loadbalancingexporter/exporter_test.go (hunk 6 above)
Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go:632]: Yes, actually line 562 does have a signal passed to the `resolverCh`. However, if we keep the old version, the resolve process will stall there, and resume when the `ConsumeTraces` in line 585 returns. In order to make the test fail for the old version and pass for the new version, I have to stop the deadlock between dns resolve and `ConsumeTraces`. do you have any better idea?

File: exporter/loadbalancingexporter/exporter_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go]: fixed

File: exporter/loadbalancingexporter/exporter_test.go (hunk 6 above)
Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go:632]: @jpkrohling could you checkout my explanation?

Comment (jpkrohling) [reply exporter/loadbalancingexporter/exporter_test.go:632]: Yes, I wanted to come up with a proposal using `WaitGroup` constructs before going out for PTO but couldn't. Would it be OK to wait for me to return? Otherwise, please open an issue and ping me there, so that we can address it later.

Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go:632]: Sure, I'm OK. as long as this pull request won't be forgotten 😬

**Comment by bogdandrutu** [comment]
@jpkrohling friendly ping, is there any comment you have left?

**Comment by jpkrohling** [comment]
From an earlier comment from this PR:

> Yes, I wanted to come up with a proposal using WaitGroup constructs before going out for PTO but couldn't. Would it be OK to wait for me to return (Jan 11)? Otherwise, please open an issue and ping me there, so that we can address it later.

**Comment by ZhengHe-MD** [comment]
@jpkrohling I don't see any changes you've made in the new branch.

File: exporter/loadbalancingexporter/exporter_test.go (hunk 6 above)
Comment (jpkrohling) [reply exporter/loadbalancingexporter/exporter_test.go:632]: This is roughly what I had in mind: https://gist.github.com/f2139cc20295c927e1455ddf67e7d11f

**Comment by jpkrohling** [comment]
It was marked automatically as stale by a GitHub bot. I added a link to a gist with what I had in mind, let me know what you think.

File: exporter/loadbalancingexporter/exporter_test.go (hunk 6 above)
Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go:632]: @jpkrohling I've read through your example. Firstly, I think there is no much difference between channels and wait groups in this specific example. and removing this line 

```go
// give limited but enough time to rolling updates. otherwise this test
// will still pass due to the 10 secs of sleep that is used to simulate
// unreachable backends.
go func() {
	time.Sleep(50 * time.Millisecond)
	resolverCh <- struct{}{}
}()
```

will make the codes before this PR pass the test either, which is not what I want. Please check it out by changing the lock and unlock statements back to where they were and run this test again, you will get the idea.

Comment (jpkrohling) [reply exporter/loadbalancingexporter/exporter_test.go:632]: Which locks do you mean? I tried commenting out the unlocks for the `updateLock` and the test failed as expected.

Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go:632]: @jpkrohling just revert the changes here

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1798/files#diff-10a4cf12dc83f1a8b5c1888cac36c0a22dc9a1710d3f5657d6d579f753367cfcL192-L196

and here

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1798/files#diff-10a4cf12dc83f1a8b5c1888cac36c0a22dc9a1710d3f5657d6d579f753367cfcR210-R216

Comment (jpkrohling) [reply exporter/loadbalancingexporter/exporter_test.go:632]: The test still failed when I reverted the mentioned changes.

Comment (ZhengHe-MD) [reply exporter/loadbalancingexporter/exporter_test.go:632]: hmm..., interesting, I do it locally again and it succeeded.

here is the `exporter_test.go`:

https://gist.github.com/ZhengHe-MD/82c9cd4db3dff6694ed2e7240f277cc5

and here is the `exporter.go`:

https://gist.github.com/ZhengHe-MD/669aaa893273dc59c33f0dc5db04e74d

Comment (jpkrohling) [reply exporter/loadbalancingexporter/exporter_test.go:632]: Sorry, you are right. I moved back the lock to the wrong place. I think this change still has a risk of introducing a flaky test if we depend on timing instead of a condition, but I'm OK to give it a try and fix the test if necessary.

**Review by jpkrohling — APPROVED** [review]
